### ISSUE TITLE
template: write_file should be configurable to honor sandbox and have information on where sandbox is

### ISSUE THREAD

**Issue by harningt** [hashicorp/nomad#14769]
### Proposal

The `writeFile` currently is blocked by default because it allows for arbitrary access to files on the filesystem.

A version that honors the sandboxing would be useful and seems like it would just leverage the `file` utility. I'm not entirely certain - but it might also be important to switch the _view_ of that operation so that `writeFile "${NOMAD_SECRETS_DIR}/file.pem"` would write to the expected place (normally /secrets/file.pem inside the view of the container).

### Use-cases

This would make it so that using writeFile could be used to address the awkward scenario that pkiCert introduces - for nomad, you must have your entire PKI definition done in one template and most tools using certificate / private keys expect to find them in separate files.

It would also be a workaround for the changed uid/gid causing rewrites - you could have your master file written as the default user w/ locked down permissions - and writeFile could be used to setup the expected parts of the files as the appropriate uid/gid/perms.

### Attempted Solutions

A workaround is to make sure that all of your clients have the same file layout and compute the expected 'secret' path (ex: some variant of `/var/nomad/alloc/${NOMAD_ALLOC}/${NOMAD_TASK}/secrets`
... while also making sure you vet all templates using `writeToFile` to make sure you don't accidentally clobber system files.

**Comment by harningt** [comment]
Issue brought up in discussion: https://discuss.hashicorp.com/t/uid-gid-ownership-of-template-output-in-nomad-secrets-dir-for-postgres/37239/6

**Comment by tgross** [comment]
Hi @harningt! Unfortunately the reason that we have `writeFile` in the default denylist is because we _can't_ sandbox it properly, or at least not in a way that doesn't break a whole bunch of existing users. This has to do with the way that Nomad implemented `template` by "embedding" consul-template as a library. Nomad prepares the environment for the template to run in and then hands it off to consul-template. Nomad doesn't know how to parse the template, and consul-template doesn't know anything about the sandbox. That made for a reasonable approach to implementation (although one that's been a source of a lot of security issues) but as you've seen it has some sharp edges that make for a pretty lousy user experience.

Our proposed solution to this is to move the entire templating process into the sandbox (ref https://github.com/hashicorp/nomad/issues/12301) so that we can let users use arbitrary functions without the risk of compromising the Nomad client agent. That's on our roadmap for the near-ish term (can't give you a timeline), and that should cover this set of problems.

I'm going to close this issue out as a wontfix because it's covered by #12301, but thanks for opening this issue! I saw the other issue you opened on unexpected re-renders and I'll be taking a look at that shortly.
